### DISCUSSION TITLE
CS9236: Compiling requires binding the lambda expression at least 100 times

### DISCUSSION THREAD

**Discussion by mu88** [dotnet/roslyn#74799] [Q&A] [answered]
Hi,

I'm now starting to see `CS9236: Compiling requires binding the lambda expression at least 100 times` for the following code:  
```csharp
private static int GetNumberOfProcessedRecordsByState(IReadOnlyCollection<BulkProcessingResult>[] bulkProcessingResults, ProcessState state) =>
    bulkProcessingResults.Sum(
        results => results
            .Where(result => result.ProcessState == state)
            .Sum(result => result.ProcessedRecords));

private static int GetNumberOfSavedRecordsByState(IReadOnlyCollection<BulkProcessingResult>[] bulkProcessingResults, ProcessState state) =>
    bulkProcessingResults.Sum(
        results => results
            .Where(result => result.ProcessState == state)
            .Sum(result => result.SavedRecords));
```

Unfortunately, I don't really find any documentation about this issue. I'd appreciate it if you guys could shed some light on it, and explain how it's triggered and how it can be resolved.

My understanding from the message would be that I should convert the code to something like this:  
```csharp
private static int GetNumberOfProcessedRecordsByState(IReadOnlyCollection<BulkProcessingResult>[] bulkProcessingResults, ProcessState state) =>
    bulkProcessingResults.Sum(
        (IReadOnlyCollection<BulkProcessingResult> results) => results
            .Where(result => result.ProcessState == state)
            .Sum((BulkProcessingResult result) => result.ProcessedRecords));

private static int GetNumberOfSavedRecordsByState(IReadOnlyCollection<BulkProcessingResult>[] bulkProcessingResults, ProcessState state) =>
    bulkProcessingResults.Sum(
        (IReadOnlyCollection<BulkProcessingResult> results) => results
            .Where(result => result.ProcessState == state)
            .Sum((BulkProcessingResult result) => result.SavedRecords));
```

However, as I don't find any docs or examples, I'd like to understand whether this is the intended fix and why it is necessary at all (some in-depth explanation would be very much appreciated).

Thank you :)

**Comment by huoyaoyuan** [comment]
The name of code 9236 is TooManyLambdas:

https://github.com/dotnet/roslyn/blob/75b26a2088e76855528483674a1fbeea0137241b/src/Compilers/CSharp/Portable/Errors/ErrorCode.cs#L2311

And the triggering code is here:

https://github.com/dotnet/roslyn/blob/75b26a2088e76855528483674a1fbeea0137241b/src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs#L446-L459

Sample in unit test is here:

https://github.com/dotnet/roslyn/blob/75b26a2088e76855528483674a1fbeea0137241b/src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs#L291-L346

The samples are about deeply nested lambda causing overload resolution to take too much times.

**Reply by mu88** [reply]
Would my second code snippet solve the issue? And if yes, why? If not, how do I fix it?

**Reply by huoyaoyuan** [reply]
As the second unit test hints, adding more explicit type annotation for lambdas would reduce the work of overload resolution, and likely resolving the issue.

**Reply by ufcpp** [reply]
```cs
    bulkProcessingResults.Sum(
        results => results
            .Where(result => result.ProcessState == state)
            .Sum(int (result) => result.ProcessedRecords)); // Use explicit return type.
```

**Reply by mu88** [reply]
Thx - do you have any in-depth explanations/docs/blogs on why this is necessary?

**Reply by CyrusNajmabadi** [reply, marked as answer]
Overload resolution with lambdas has to try each possible conversion from each lambda to each target type.  This then involves binding the bodies of the lambdas, which may involves doing the same in the interior, etc etc.  This can trivially lead to a combinatorial explosion which can add up to seconds or minutes of compilation time. 

By supplying the argument and return types, the compiler does not have to do this work, massively speeding things up in some pathological cases.

**Reply by rawling** [reply]
Has this got worse recently due to improvements (= more options) in overload resolutions for lambdas or something? I've never had an issue before but today have noticed VS (Intellisense) slowing down when I don't address these messages.
(Edit: only seems to dislike `Sum`, so something related to that?)

**Reply by CyrusNajmabadi** [reply]
It's possible you're compiling against a library that added more overloads. This would cause more binding. 

> (Edit: only seems to dislike Sum, so something related to that?)

It's likely there are more Sum overloads now.

**Reply by CyrusNajmabadi** [reply]
> have noticed VS (Intellisense) slowing down when I don't address these messages.

Yes. This will slow down intellisense. We have to bind to understand what your code means. If binding needs to try hundreds of thousands of combinations, then that will slow down intellisense. Either break up the expression into smaller pieces, or provide typing hints.

**Reply by rawling** [reply]
> It's likely there are more Sum overloads now.

I think maybe a combination of this, and that I've quite possibly never triple-nested `Sum` calls before (thanks, Advent of Code!)

> or provide typing hints

Yeah - looks like I can fix it up by just specifying the return type on each of the lambdas that I pass into `Sum`s. Useful new feature to have!

Thanks for following up.

**Reply by 333fred** [reply]
> and that I've quite possibly never triple-nested Sum calls before

This is the reason. Type inference is exponential in complexity. Triple nesting with generic lambdas is an excellent way to stress test your CPU.

**Reply by ufcpp** [reply]
Could it be solved using OverloadResolutionPriority?, although This is more of a [runtime](https://github.com/dotnet/runtime/) issue than csharplang.

https://gist.github.com/ufcpp/0b4a77b8e129eb79a47c237719b602cf

**Reply by 333fred** [reply]
Not really. ORPA is the last step in overload resolution. Type inference already had to happen to determine applicability.

**Reply by ufcpp** [reply]
[The example code](https://gist.github.com/ufcpp/0b4a77b8e129eb79a47c237719b602cf) actually stops the CS9236 warnings, but does it still slow down the compilation?

**Reply by 333fred** [reply]
It will still slow down the compilation. ORPA does not stop lambda inference from occurring.

**Reply by ufcpp** [reply]
Then, shouldn't CS9326 be produced in the example as well?

**Reply by 333fred** [reply]
I suppose it's possible that it could have an effect in some particularly degenerate edge cases, but I do not expect it to have a general impact on performance.
